engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -183,6 +184,7 @@ Text toDisplayString(boolean allowSideEffects, @CachedLibrary("this") InteropLib
     try {
       return TypesGen.expectText(atoms.invokeMember(this, "to_text"));
     } catch (UnsupportedMessageException
+        | AbstractTruffleException
```

Comment:
Ok, so we catch the error here - so we _do_ have point where we see that `to_text` failed, right? Cannot we then throw a Panic exception from within here instead of returning the fallback error?

Reply:
We can throw an exception, but it ends up at [truffle catch block](https://github.com/oracle/graal/blob/83c37dde9922be071b8d6efe3b0a140bbb1231f2/truffle/src/com.oracle.truffle.polyglot/src/com/oracle/truffle/polyglot/PolyglotWrapper.java#L181) - there is not much to do about it.

Reply:
Ah, I see! Well that's too bad. In that case, I guess the current solution is right.

Reply:
However, I feel like it could be good to at least signalize that the default `to_text` method has failed to execute and we are getting a replacement instead - personally I'm very much against returning such replacements without telling the user.

Could we maybe change the replacement to to something like
```
[Error in method to_text of <fallback-toString>: <Exception message>]
```
?

I think this could be the best middle ground - we still return some valid text representation, we still show the fallback representation so that the user can know what was the object that failed - and when we include the exception message too, the user can also have some clue as to what has happened (of course it's not as good as with a stacktrace, but including the stack trace would make the representation too long, so I think just the message is a reasonable middle ground).

Reply:
So I'd suggest to amend the 2nd and 3rd tests to be:

```
assertEquals("[[[Error in method `to_text` of [Cb y]: Illegal_State.Error "B"]]]", v.asString());
```
and
```
assertEquals("[[[Panic in method `to_text` of [Cc z]: Illegal_State.Error "C"]]]", v.asString());
```
respectively.

Ideally we should distinguish a dataflow error (writing `Error in method` or even `Dataflow error in method`) and panics (`Panic in method`). We could also use `Exception in method` if it was a Java exception, but I guess since Java exceptions are the same as Panics for Enso, I think just Panic is OK.

What do you think? Does it sound OK?

Reply:
5fa1b5c98d2615ad567ad6a3c4da44bd129f4229 is the closest I could get.